tests/unit/samplegen/test_samplegen.py

```diff
@@ -12,453 +12,670 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import pytest
 import yaml
+import itertools
+import pytest
+
+from typing import TypeVar
 from collections import namedtuple
-from textwrap import dedent
 
-import gapic.samplegen.samplegen as samplegen
+import gapic.schema.wrappers as wrappers
 import gapic.samplegen.yaml as gapic_yaml
+import gapic.samplegen.samplegen as samplegen
 
-
+from common_types import DummyField, DummyMessage, DummyMethod
 from gapic.samplegen import utils
 
+
+def message_factory(exp: str, repeated_iter=itertools.repeat(False)) -> DummyMessage:
+    # This mimics the structure of MessageType in the wrappers module:
+    # A MessageType has a map from field names to Fields,
+    # and a Field has an (optional) MessageType.
+    # The 'exp' parameter is a dotted attribute expression
+    # used to describe the field and type hierarchy,
+    # e.g. "mollusc.cephalopod.coleoid"
+    toks = exp.split(".")
+    messages = [DummyMessage({}, tok.upper() + "_TYPE") for tok in toks]
+    for base, field, attr_name, repeated_field in zip(
+        messages, messages[1:], toks[1:], repeated_iter
+    ):
+        base.fields[attr_name] = DummyField(field, repeated=repeated_field)
+
+    return messages[0]
+
+
 # validate_response tests
 
 
 def test_define():
     define = {"define": "squid=$resp"}
-    samplegen.Validator().validate_response([define])
+
+    samplegen.Validator(
+        DummyMethod(output=message_factory("mollusc"))
+    ).validate_response([define])
 
 
 def test_define_undefined_var():
     define = {"define": "squid=humboldt"}
     with pytest.raises(samplegen.UndefinedVariableReference):
-        samplegen.Validator().validate_response([define])
+        samplegen.Validator(
+            DummyMethod(output=message_factory("mollusc"))
+        ).validate_response([define])
 
 
 def test_define_reserved_varname():
     define = {"define": "class=$resp"}
     with pytest.raises(samplegen.ReservedVariableName):
-        samplegen.Validator().validate_response([define])
+        samplegen.Validator(
+            DummyMethod(output=message_factory("mollusc"))
+        ).validate_response([define])
 
 
 def test_define_add_var():
-    samplegen.Validator().validate_response([{"define": "squid=$resp"},
-                                             {"define": "name=squid.name"}])
+    samplegen.Validator(
+        DummyMethod(output=message_factory("mollusc.name"))
+    ).validate_response([{"define": "squid=$resp"}, {"define": "name=squid.name"}])
```

Comment:
Minor: I suggest consistency in whether you have the `response` inline or via a separate variable (cf previous cases above)

Reply:
I'll clean it up in a followup review.